File: normal_order_python/236.LowestCommonAncestorofaBinaryTree.py
'''


'''



# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
	# much faster
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if not root:
            return None
        stack = [root] 
        parent = {root:None}
        while stack:
            cur = stack.pop()
            if cur.right:
                stack.append(cur.right)
                parent[cur.right] = cur
            if cur.left:
                stack.append(cur.left)
                parent[cur.left] = cur
            if p in parent and q in parent:
                break
        # print(parent)
        ancestor = set()
        while p:
            ancestor.add(p)
            p = parent[p]
        while q not in ancestor:
            q = parent[q]
        return q

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if not root:
            return None
        stack = [(root, [root])] # stack stored with current node and path
        two = {}
        while stack:
            cur, path = stack.pop()
            if cur == p:
                # print(path)
                two[p] = path+[cur]
            if cur == q:
                # print(path)
                two[q] = path+[cur]
            if len(two) == 2:
                break
            if cur.right:
                stack.append((cur.right, path+[cur]))
            if cur.left:
                stack.append((cur.left, path+[cur]))
        # print(two[p])
        return [x for x in two[p] if x in two[q]][-1]

File: normal_order_java/236.LowestCommonAncestorOfaBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    // faster and smarter
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null || root == p || root == q) return root;
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        if(left != null && right != null) return root;
        return left != null? left: right;
        // left == null && right == null return null;
        // left == null && right != null return right;
        // left != null && right == null return left;
    }
}


class Solution {
    TreeNode ans;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        Map<TreeNode, TreeNode> parent = new HashMap<TreeNode, TreeNode>();
        Set<TreeNode> set = new HashSet<TreeNode>();
        dfs(parent, set, root, p, q);
        return ans;
    }
    
    public void dfs(Map<TreeNode, TreeNode> parent, Set<TreeNode> set, TreeNode root, TreeNode p, TreeNode q){
        if(set.size() == 2 || root == null) return;
        if(root == p || root == q){
            set.add(root);
            if(set.size() == 2){
                Set<TreeNode> candi = new HashSet();
                while(parent.containsKey(p)){
                    // System.out.println(p.val);
                    candi.add(p);
                    p = parent.get(p);
                }
                candi.add(p);
                while(!candi.contains(q)){
                    q = parent.get(q);
                }
                this.ans = q;
                return;
            }
        }
        if(root.left != null){
            parent.put(root.left, root);
            // System.out.println(root.left.val + " " + 1);
            dfs(parent, set, root.left, p, q);
        }
        if(root.right != null){
            // System.out.println(root.right.val + " " + 2);
            parent.put(root.right, root);
            dfs(parent, set, root.right, p, q);
        }
        return;
    }
}